### DISCUSSION TITLE
Android SSL certificate stopped working after Android 13 upgrade

### DISCUSSION THREAD

**Discussion by CwithW** [mitmproxy/mitmproxy#6460] [Q&A] [answered]
#### Problem Description

I am using the magisk module generated by mitmproxy to install SSL certificate into the Android system certificate store and capture SSL traffic. It works until I recently upgrades my phone's Android OS then it stopped working. 

The mitmproxy certificate can be seen installed into the phone and in the system certificate list, but apps are no longer trusting the certificate.

The logs given by Chrome browser:
```
NET::ERR_CERT_AUTHORITY_INVALID
Subject: www.baidu.com

Issuer: mitmproxy

Expires on: 2024年11月5日

Current date: 2023年11月5日

PEM encoded chain:
-----BEGIN CERTIFICATE-----
MIIDRTCCAi2gAwIBAgIUchepgKnGAI21lrSwYhAwMwbFhkQwDQYJKoZIhvcNAQEL
BQAwKDESMBAGA1UEAwwJbWl0bXByb3h5MRIwEAYDVQQKDAltaXRtcHJveHkwHhcN
MjMxMTAzMTY0MTM4WhcNMjQxMTA0MTY0MTM4WjAYMRYwFAYDVQQDDA13d3cuYmFp
ZHUuY29tMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAjY5+aHlj7r6g
lZs1pUfpkPBuE6eOIyvbv+bv3bOvaYJ7ZQ5ONnDiTHnpNUjIYbVgzlc/tveZv/Wx
a6qfKeSlT1jgx+gzV4O4q9KDyNielEmID4sE7CYZ0WkGOhotn4UMQj9QGP3XR/Xr
yal4AUPsNoRBdLI6k46JIHVDKCbmBCF3OAK2cRvOnva4cHOGTTOFYn2dxyd9XRcr
KP/zX3wMP7jZxUx6wIshGPuQCEF6xsvKTvfuv5BFRmucUbsHJ/oJJJCZ+uj5eZSP
f740MH5kylpef/Ct9bLYdoPoR46DNIAt7lecFVGV6XK1goZIXsjvuu8Ye8d0qrgr
9GBb8Sj7uwIDAQABo3cwdTATBgNVHSUEDDAKBggrBgEFBQcDATAeBgNVHREEFzAV
gg13d3cuYmFpZHUuY29thwSZA+5uMB0GA1UdDgQWBBR4GuXPad/PmUSQg3//Fip/
ggvejDAfBgNVHSMEGDAWgBR4GuXPad/PmUSQg3//Fip/ggvejDANBgkqhkiG9w0B
AQsFAAOCAQEAax+Kze0OviTbFRZoolxBBdOwKcYBJd1ErelzX5NSxxrcI4hFsV0b
1hLu211USsxFrFUI8c3kPkVajzFjm7JMbE7s6CN0dZjJlYS4/sDKO9FrBrdTadw+
x9pGEAQCIlfaaWGKp3juWyz9eF9W2wIR9eamYBqDiIwYqSglc5HTp8I78OsWUEf8
lWk3bRmi+cG627kZb0Isb7kjXyaqUGlKk4wbkUMEiKgGagXhy2xjeKt+KED9v4FF
DWwIRKzwx5pvvwlTkyY2JX3pad2qHOZjyHVru0Zc59ECJ5L3s40XicZHIfXoXfmA
RA56x7VTJeLXdE3BNUx05uz7dVuYxMDbUg==
-----END CERTIFICATE-----
```

The logs in mitmproxy:
```
[16:42:20.897][10.42.0.233:49930] server connect 153.3.238.110:443
[16:42:20.990][10.42.0.233:49914] Client TLS handshake failed. The client does not trust the proxy's certificate for www.baidu.com (OpenSSL Error([('SSL routines', '', 'sslv3 alert certificate unknown')]))
```

The upgraded Android OS is MIUI14, Android13 with Android security update 2023-10-01.

The same magisk module and certificate is working on my other phone with Android 10.

#### Steps to reproduce the behavior:
1.  Install the latest Android upgrade
2.  Used mitmweb with transparent proxy `mitmweb --web-host 0 --mode transparent --showhost --set stream_large_bodies=10m` and setup iptables forwarding
3.  Access https website with Chrome

#### System Information
Paste the output of "mitmproxy --version" here.

```
Mitmproxy: 10.1.3 binary
Python:    3.12.0
OpenSSL:   OpenSSL 3.1.4 24 Oct 2023
Platform:  Linux-5.15.0-86-generic-x86_64-with-glibc2.35
```

**Comment by ndoulgeridis** [comment]
Are there any news for Android 13 and this issue? Any workaround we can do to make it work? Except cannot be used on Chrome, any app also cannot access Internet:

`java.security.cert.CertPathValidatorException: Trust anchor for certification path not found`

On mint console I get:

`Client TLS handshake failed. The client does not trust the proxy's certificate for accounts.google.com (OpenSSL Error([('SSL routines', '', 'sslv3 alert certificate unknown')]))`

**Comment by CwithW** [comment]
Still unsolved now, Android 14 requires you to throw the certificate into `/apex/com.android.conscrypt/cacerts` and do bind mount ( as seen in [https://github.com/nccgroup/ConscryptTrustUserCerts](https://github.com/nccgroup/ConscryptTrustUserCerts) ), but strangely, I am on Android 13 and I do not have the `/apex/com.android.conscrypt/cacerts` folder.

**Comment by CwithW** [comment]
on further investgation, the process which the browser(mark.via in this case) is in cannot see the ca cert, but the zygote process, zygote64 process and adb process can see the ca cert. this is strange considering all Android process is a fork of zygote.
```shell
rubens:/ # ps -ef | grep -v grep | grep 10378
u0_a414      10378   856 6 21:40:54 ?     00:00:13 mark.via
rubens:/ # nsenter --mount=/proc/10378/ns/mnt -- ls -lat /system/etc/security/cacerts  | grep 2024
1|rubens:/ # nsenter --mount=/proc/self/ns/mnt -- ls -lat /system/etc/security/cacerts  | grep 2024
drwxr-xr-x 2 root root 2580 2024-03-05 21:39 .
-rw-r--r-- 1 root root  935 2024-03-05 21:39 e5c3944b.0
rubens:/ # nsenter --mount=/proc/$(pidof zygote)/ns/mnt -- ls -lat /system/etc/security/cacerts  | grep 2024
drwxr-xr-x 2 root root 2580 2024-03-05 21:39 .
-rw-r--r-- 1 root root  935 2024-03-05 21:39 e5c3944b.0
rubens:/ # nsenter --mount=/proc/$(pidof zygote64)/ns/mnt -- ls -lat /system/etc/security/cacerts  | grep 2024
drwxr-xr-x 2 root root 2580 2024-03-05 21:39 .
-rw-r--r-- 1 root root  935 2024-03-05 21:39 e5c3944b.0
```

**Comment by CwithW** [comment, marked as answer]
Try removing your target app in Magisk Exclusion List and disable all magisk hiding modules(eg. Shamiko, Magisk Delta, Magisk Hide in old versions of Magisk) first. If you still cannot use the CA cert or unwilling to disable magisk hiding module because your target app is detecting magisk, see below.

Some magisk hiding modules actively unmount bind mounts created by Magisk to hide the trace of Magisk. Somehow recent versions of Shamiko will unmount changes in `/system/etc/security` and cause the cert to disappear for apps in Magisk Exclusion List.

You can make Shamiko and CA cert co-exist by manually bind mounting the cert yourself within the mount namespace of the apps.

I wrote a script that does bind mount on `/system/etc/security/cacerts` for zygote and zygote64 and every apps again, this works for me, but I need to execute this on every Android boot and before opening any Android app.
note: you need to put your ca cert into `/system/etc/security/cacerts`(via magisk) before running this. also run as root

```shell

#!/system/bin/sh

# Workaround for Android 13 ca cert not working
# copy and bind mounts /system/etc/security/cacerts for zygote, zygote64 and every Android app
# I don't know why this is needed, but this works
# You only need this on some version of Android 13 that simply putting certificate into /system/etc/security/cacerts does not work
# For Android 14, see https://github.com/nccgroup/ConscryptTrustUserCert
# Don't use this script with ConscryptTrustUserCert magisk module at the same time, because both of them used `/data/local/tmp/tmp-ca-copy` path.

# note: you need to put your ca cert into `/system/etc/security/cacerts`(via magisk) before running this. also run as root

if [ $(id -u) -ne 0 ]; then
    echo "This script must be run as root" 1>&2
    exit 1
fi

set -e
set -x
rm -rf /data/local/tmp/tmp-ca-copy || true;
mkdir -p -m 700 /data/local/tmp/tmp-ca-copy;
cp /system/etc/security/cacerts/* /data/local/tmp/tmp-ca-copy/;
ZYGOTE_PID=$(pidof zygote || true);ZYGOTE64_PID=$(pidof zygote64 || true);
APP_PIDS=$(echo "$ZYGOTE_PID $ZYGOTE64_PID" | \xargs -n1 ps -o 'PID' -P | \grep -v PID);
for Z_PID in $ZYGOTE_PID; do
    if [ -n "$Z_PID" ]; then
        /system/bin/nsenter --mount=/proc/$Z_PID/ns/mnt -- /bin/mount --bind /data/local/tmp/tmp-ca-copy /system/etc/security/cacerts;
    fi
done
for Z_PID in $ZYGOTE64_PID; do
    if [ -n "$Z_PID" ]; then
    /system/bin/nsenter --mount=/proc/$Z_PID/ns/mnt -- /bin/mount --bind /data/local/tmp/tmp-ca-copy /system/etc/security/cacerts;    
    fi
done
for PID in $APP_PIDS; do
    /system/bin/nsenter --mount=/proc/$PID/ns/mnt -- /bin/mount --bind /data/local/tmp/tmp-ca-copy /system/etc/security/cacerts;
done;


```

**Reply by shenghuang147** [reply]
@CwithW 

You can refer to [Certificate injection](https://www.lategege.com/?p=1452)

**Reply by Hamza-Khiar** [reply]
where to put this script tho??
do i need to put it in magisk path /data/adb/modules path as a magisk module

**Reply by shenghuang147** [reply]
@Hamza-Khiar You can place the script in the `/data/adb/post-fs-data.d` directory and give it executable permissions so that it will be automatically loaded on every boot.

**Reply by Hamza-Khiar** [reply]
@CwithW I tried this and didn't work for one reason, so basically i am inserting the certs with magisk module, but when putting the shell script above in `post-fs-data.d` what happens is it gets executed before magisk mounts/overlays your crts in /system/etc/security/cacerts.

i checked the /tmp dir i didn't find my cert in there.

the solution is to either put it in services.d so that it gets executed as a `late_start` script; 

https://topjohnwu.github.io/Magisk/guides.html#boot-scripts

https://topjohnwu.github.io/Magisk/details.html#paths-in-data

the workaround works.

**Reply by fireattack** [reply]
> @CwithW I tried this and didn't work for one reason, so basically i am inserting the certs with magisk module, but when putting the shell script above in `post-fs-data.d` what happens is it gets executed before magisk mounts/overlays your crts in /system/etc/security/cacerts.
> 
> i checked the /tmp dir i didn't find my cert in there.
> 
> the solution is to either put it in services.d so that it gets executed as a `late_start` script;
> 
> https://topjohnwu.github.io/Magisk/guides.html#boot-scripts
> 
> https://topjohnwu.github.io/Magisk/details.html#paths-in-data
> 
> the workaround works.

I put the script in `/data/adb/service.d` but it still doesn't work for a Pixel7/Android 13 AVD :(
